src/transformer_traits.jl

```diff
@@ -0,0 +1,421 @@
+# tuple-nesting in the returned value to hash. This ensures that if we want two types to
+# transform to the same string, the hashed value doesn't depend on how many transformations
+# deep we go to "find" this identical string (unless there is distinct structure that *must*
+# be hashed give its `StructType`).
+pair_structure(x, ::Nothing) = x
+pair_structure(x, y) = (x, y)
+function transformer(::Type{T}, context::TypeHashContext) where {T<:Type}
+    return Transformer(T -> pair_structure(transform_type(T, parent_context(context)),
+                                           internal_type_structure(T, StructType_(T))))
+end
+@inline StructType_(T) = StructType(T)
+StructType_(::Type{Union{}}) = StructTypes.NoStructType()
+
+# NOTE: `internal_type_structure` implements mandatory elements of a type's structure that
+# are always included in the hash; this ensures that the invariants required by type
+# hoisting hold
+internal_type_structure(T, trait) = nothing
+
+#####
+##### Hashing Types as Values
+#####
+
+struct TypeAsValue <: StructTypes.StructType end
+hash_trait(::Type) = TypeAsValue()
+
+struct TypeAsValueContext{T}
+    parent::T
+end
+parent_context(x::TypeAsValueContext) = x.parent
+
+function hash_type!(hash_state, context, ::Type{<:Type})
+    return update_hash!(hash_state, "Base.Type", context)
+end
+# these methods are required to avoid method ambiguities
+function hash_type!(hash_state, context::TypeHashContext, ::Type{<:Type})
+    return update_hash!(hash_state, "Base.Type", context)
+end
+function hash_type!(hash_state, context::TypeAsValueContext, ::Type{<:Type})
+    return update_hash!(hash_state, "Base.Type", context)
+end
+
+function transformer(::Type{<:Type}, context::TypeAsValueContext)
+    return Transformer(T -> pair_structure(transform_type_value(T, context),
+                                           internal_type_structure(T, StructType_(T))))
+end
+
+hash_type!(hash_state, ::TypeAsValueContext, ::Type) = hash_state
+function stable_hash_helper(::Type{T}, hash_state, context, ::TypeAsValue) where {T}
+    type_context = TypeAsValueContext(context)
+    transform = transformer(typeof(T), type_context)::Transformer
+    tT = transform(T)
+    return stable_hash_helper(tT, hash_state, type_context, hash_trait(transform, tT))
+end
+
+#####
+##### Function Hashes
+#####
+
+# remember: functions can have fields; in general StructTypes doesn't assume these are
+# serialized but here we want that to happen by default, so e.g. ==(2) will properly hash
+# both the name of `==` and `2`.
+hash_trait(::Function) = StructTypes.UnorderedStruct()
+
+transform_type(::Type{T}) where {T<:Function} = nameof_string(T)
+
+#####
+##### DataType
+#####
+
+transform_type_by_trait(::Type{T}, ::StructTypes.DataType) where {T} = nameof_string(T)
+
+sorted_field_names(T::Type) = TupleTools.sort(fieldnames(T); by=string)
+@generated function sorted_field_names(T)
+    return TupleTools.sort(fieldnames(T); by=string)
+end
```

Comment:
This one we will probably need.

To me this seems like the least problematic of the `@generated` functions, since it is using a function defined in a separate package (so issues around the ordering of this function re the rest of the package aren't an issue) and it is a relatively straightforward transformation of the input.

I don't see an obvious way around it either: we could use an approach storing results in an IdDict, though based on the earlier benchmarks, my guess is that this would be a substantial hit to speed — though perhaps just directly sorting wouldn't be too slow??